### PR TITLE
Endpoint that allows putting app data without knowing the appDataHash

### PR DESCRIPTION
**Pull request by fleupold**
# Description
One unnecessary annoyance about the current API is that in order to upload an AppData pre-image, one has to already know its hash (which can be computed using the CoW SDK but may be hard to get if one is using a different language or does not want to depend on the SDK).

This PR introduces an endpoint that allows uploading appData without first having to compute the hash. Instead the hash is returned in the response.

# Changes

- [x] Make `AppDataHash` optional in existing route
- [x] Document new route

## How to test
```bash
curl -X 'PUT' \
  'https://api.cow.fi/mainnet/api/v1/app_data' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "fullAppData": "{\"version\":\"0.9.0\",\"metadata\":{}}"
}'
```

### CODE REVIEW

**Comment by MartinquaXD** [comment]
Would it be possible to rather make the `hash` value optional in the request?
This seems to be in line with the `from` field in order creation requests where we only check that the signature matches the owner if that value is provided. If the value is missing we assume everything is correct and continue.

**Comment by fleupold** [comment]
> Would it be possible to rather make the hash value optional in the request?

My worry is that this would not be backwards compatible and make this change much harder to get adopted (requires updating all frontends, SDKs, etc)

**Review by squadgazzz — APPROVED** [review]

File: crates/orderbook/src/app_data.rs [hunk 1]
```diff
@@ -42,27 +42,29 @@ impl Registry {
...
         let validated = self
             .validator
             .validate(document)
             .map_err(RegisterError::Invalid)?;
-        if hash != validated.hash {
-            return Err(RegisterError::HashMismatch {
-                expected: hash,
-                computed: validated.hash,
-            });
+        if let Some(hash) = hash {
```
Comment (MartinquaXD) [inline crates/orderbook/src/app_data.rs:52]: Could lose 1 level of indentation with:
```rust
if hash.is_some_and(|hash| hash != validated.hash) {
```

File: crates/orderbook/src/api/put_app_data.rs [hunk 2]
```diff
@@ -10,19 +10,22 @@ use {
 
 fn request(
     max_size: usize,
-) -> impl Filter<Extract = (AppDataHash, AppDataDocument), Error = Rejection> + Clone {
-    warp::path!("v1" / "app_data" / AppDataHash)
+) -> impl Filter<Extract = (Option<AppDataHash>, AppDataDocument), Error = Rejection> + Clone {
+    let opt = warp::path::param::<AppDataHash>()
```
Comment (MartinquaXD) [inline crates/orderbook/src/api/put_app_data.rs:14]: Optionally fetching the appdata hash looks correct but would be nice to add a test.

**Review by MartinquaXD — APPROVED** [review]

**Pull request merged**